controllers/metal3.io/bmceventsubscription_controller_test.go

```diff
@@ -0,0 +1,138 @@
+	return newBMCTestReconcilerWithFixture(&fix, initObjs...)
+}
+
+func newBMCRequest(subscription *metal3api.BMCEventSubscription) ctrl.Request {
+	namespacedName := types.NamespacedName{
+		Namespace: subscription.ObjectMeta.Namespace,
+		Name:      subscription.ObjectMeta.Name,
+	}
+	return ctrl.Request{NamespacedName: namespacedName}
+}
+
+func newSubscription(name string, spec *metal3api.BMCEventSubscriptionSpec) *metal3api.BMCEventSubscription {
+	return &metal3api.BMCEventSubscription{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "BareMetalHost",
+			APIVersion: "metal3.io/v1alpha1",
+		}, ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: *spec,
+	}
+}
+
+func newDefaultNamedSubscription(t *testing.T, name string) *metal3api.BMCEventSubscription {
+	t.Helper()
+	spec := &metal3api.BMCEventSubscriptionSpec{
+		HostName:    t.Name(),
+		Destination: "user destination",
+		Context:     "user context",
+		HTTPHeadersRef: &corev1.SecretReference{
+			Name:      defaultSecretName,
+			Namespace: namespace,
+		},
+	}
+	t.Logf("newNamedSubscription(%s)", name)
+	subscription := newSubscription(name, spec)
+	return subscription
+}
+
+func newDefaultSubscription(t *testing.T) *metal3api.BMCEventSubscription {
+	t.Helper()
+	return newDefaultNamedSubscription(t, t.Name())
+}
+
+func tryBMCReconcile(t *testing.T, r *BMCEventSubscriptionReconciler, subscription *metal3api.BMCEventSubscription, isDone BMCDoneFunc) {
+	t.Helper()
+	request := newBMCRequest(subscription)
+
+	for i := 0; ; i++ {
+		t.Logf("tryReconcile: top of loop %d", i)
+		if i >= 25 {
+			t.Fatal("Exceeded 25 iterations")
+		}
+
+		result, err := r.Reconcile(context.Background(), request)
+
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// The FakeClient keeps a copy of the object we update, so we
+		// need to replace the one we have with the updated data in
+		// order to test it. In case it was not found, let's set it to nil
+		updatedSubscription := &metal3api.BMCEventSubscription{}
+		if err = r.Get(context.TODO(), request.NamespacedName, updatedSubscription); k8serrors.IsNotFound(err) {
+			subscription = nil
+		} else {
+			updatedSubscription.DeepCopyInto(subscription)
+		}
+
+		if isDone(subscription, result) {
+			t.Logf("tryReconcile: loop done %d", i)
+			break
+		}
+
+		t.Logf("tryReconcile: loop bottom %d result=%v", i, result)
+		if !result.Requeue && result.RequeueAfter == 0 {
+			t.Fatalf("Ended reconcile at iteration %d without test condition being true", i)
+		}
+	}
+}
+
+func TestBMCAddFinalizers(t *testing.T) {
+	host := newDefaultHost(t)
+	subscription := newDefaultSubscription(t)
+	r := newBMCTestReconciler(subscription, host)
+	tryBMCReconcile(t, r, subscription,
+		func(subscription *metal3api.BMCEventSubscription, result reconcile.Result) bool {
+			t.Logf("subscription finalizers: %v", subscription.Finalizers)
+			return utils.StringInList(subscription.Finalizers, metal3api.BMCEventSubscriptionFinalizer)
+		})
```

Comment:
What do you think about simplifying this so that we call `addFinalizers` directly instead of going through the reconcile? It can be nice to test `Reconcile` properly like this but I think it will be hard to do it for more advanced cases. For unit tests I think it would be ok and maybe even preferable to just test each function separately.